### DISCUSSION TITLE
Idea: add lease mileage limit with start/end dates

### DISCUSSION THREAD

**Discussion by SDekkers** [teslamate-org/teslamate#3750] [Ideas]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

As many Tesla drivers, at least in Netherlands, are leasing the car with a contract for X years and Y km per year, it'd be nice to add the max odometer value for today in the overview, together with the remaining km "balance" left on the car for today, to quickly see if you're driving too much or can go on a long road trip without causing extra costs or having to reduce driving later on.

I've copied the main dashboard and edited it a bit to include these values based on hard-coded dates. 
![image](https://github.com/teslamate-org/teslamate/assets/913981/c4dfb4a2-c750-4640-aadf-6bc188119bd7)

The idea here is to have some extra input fields in car settings, lease start date, lease end date and yearly contract limit, in my case 35000km. We can then calculate the number of years in the start/end period, calculate the max odometer value at the end of the contract and average of km per day allowed to show the max/remaining km values.

For max mileage I've used the below query (hardcoded values as that's the easiest to implement)
```
select $__time(date), ROUND((DATE_PART('day', NOW()::TIMESTAMP - '2024-03-14 00:00:00'::TIMESTAMP) / 1826) * 175000)  as "odometer_$length_unit"
from positions 
where car_id = $car_id 
order by date desc 
limit 1;
```

For remaining mileage:
```
select $__time(date), (ROUND((DATE_PART('day', NOW()::TIMESTAMP - '2024-03-14 00:00:00'::TIMESTAMP) / 1826) * 175000) - convert_km(odometer::numeric, '$length_unit'))  as "odometer_$length_unit"
from positions 
where car_id = $car_id 
order by date desc 
limit 1;
```

In the examples I've used 1826 as number of days between March 14 2024 - March 14 2029 and 175000 as the agreed 5 year * 35000 yearly kms

If this is something that would be merged, I'm happy to build code for this and submit a PR

### Relevant log output

```shell
not applicable
```

### Type of installation

Docker

### Version

v1.28.4

**Comment by cwanja** [comment]
Great idea. I am moving this to Discussions/Ideas as it is not an "issue" in GitHub sense. 

> If this is something that would be merged, I'm happy to build code for this and submit a PR

You would need to consider how to hide these for those without lease vehicles. Maybe a flag in teslamate/settings for a lease that then shows end mileage. Grafana tiles would then be shown or hidden based upon the settings flag.

Edit: rereading your post, I see you thought about that 😊 give the code build a go and it will be reviewed!

**Comment by DrMichael** [comment]
> You would need to consider how to hide these for those without lease vehicles.

Yes, please, that was what I was thinking. I don't think, it is good to have info in the e.g. dashboard, which is irrelevant for most users. What about a separate leasing dashboard?

**Comment by netdata-be** [comment]
I'm liking the idea but how is this different from the "Extrapolated monthly mileage" and "Extrapolated annual mileage"?

This are values I use to answer the same question: Do I still have remaining milage this year / Month.

And thanks to this values I can even go OVER the limit and next month be a bit conservative and lower the extrapolated milage.

What is missing are fixed time ranges for that dashboard that could be related to the lease parameters.

**Comment by sbolten** [comment]
I have also created my own "Leasing Mileage Tracker".

The Lease Deviation (in the example below it is -398km) tells you how much your car's actual mileage differs from the mileage your lease agreement expected it to have at this point in time.  It's essentially a measure of whether you're driving more or less than your lease allows.

Or in other words: A Lease Deviation of -398km means you've driven 398 kilometers less than your lease agreement projected you should have driven by this point.  So, you're currently under your expected mileage.  This is generally a good thing, as it gives you a buffer for the remaining lease term.


![image](https://github.com/user-attachments/assets/8c1c6950-8307-4e05-825f-a806d0bca24a)

You need to create two dashboard level variables:

- Text box: km_per_year
- Text box: projection_years

Query for the chart is the following:

```
WITH actual AS (
    SELECT
        date_trunc('day', date) AS "time",
        MAX(odometer) AS odometer
    FROM positions
    WHERE car_id = ${car_id}
    GROUP BY date_trunc('day', date)
),
projection AS (
    SELECT 
    gs AS "time",
    (EXTRACT(EPOCH FROM (gs - $__timeFrom()::date)) / (${projection_years} * 365 * 24 * 60 * 60)) * ${km_per_year}*${projection_years} AS projected_km
FROM generate_series(
    $__timeFrom()::date::timestamp,
    $__timeFrom()::date::timestamp + make_interval(years => ${projection_years}),
    '1 day'::interval
) AS gs
)
SELECT 
    a."time"::date,
    a.odometer as "Odometer (km)",
    p.projected_km as "Lease Allowance (km)",
    (a.odometer - p.projected_km) AS "Lease Deviation (km)"
FROM actual a
JOIN projection p ON a."time" = p."time"
ORDER BY a."time" DESC;
```

Query for the stat panel is this:

```
WITH actual AS (
    SELECT
        MAX(odometer) AS odometer
    FROM positions
    WHERE car_id = ${car_id}
),
projection AS (
    SELECT 
        (EXTRACT(EPOCH FROM (gs - $__timeFrom()::date)) / (${projection_years} * 365 * 24 * 60 * 60)) * ${km_per_year}*${projection_years} AS projected_km
    FROM generate_series(
        $__timeFrom()::date::timestamp,
        $__timeFrom()::date::timestamp + make_interval(years => ${projection_years}),
        '1 day'::interval
    ) AS gs
    WHERE gs::date = CURRENT_DATE
)
SELECT 
    a.odometer as "Odometer (km)",
    p.projected_km as "Lease Allowance (km)",
    (a.odometer - p.projected_km) AS "Lease Deviation (km)"
    FROM actual a, projection p; 
```
